### PR TITLE
bpo-46110: Add a recursion check to avoid stack overflow in the PEG parser

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue46110

### CODE REVIEW

File: Tools/peg_generator/pegen/c_generator.py [hunk 1]
```diff
@@ -37,6 +37,8 @@
 #  define D(x)
 #endif
 
+# define MAXSTACK 1700
```
Comment (pablogsal) [inline Tools/peg_generator/pegen/c_generator.py]: This was taken from the old stack limit:

https://github.com/python/cpython/blob/8202a7e3dd55d551906390742fc748d0747256b8/Parser/parser.h#L10

File: Lib/test/test_syntax.py [hunk 2]
```diff
@@ -1729,6 +1729,13 @@ def test_syntax_error_on_deeply_nested_blocks(self):
 """
         self._check_error(source, "too many statically nested blocks")
 
+    @support.cpython_only
+    def test_error_on_parser_stack_overflow(self):
+        source = "-" * 2000 + "4"
+        for mode in ["exec", "eval", "single"]:
+            with self.assertRaises(MemoryError):
+                compile(source, "<string>", mode)
```
Comment (isidentical) [inline Lib/test/test_syntax.py]:
```suggestion
            with self.subTest(mode=mode):
                with self.assertRaises(MemoryError):
                    compile(source, "<string>", mode)
```

**Review by isidentical — APPROVED** [review]
LGTM

**Comment by cmcmarrow** [comment]
thanks for the PR!

**Comment by pablogsal** [comment]
Unfortunately, I am not still convinced about this since 1700 frames are reached easily with 60 or so parentheses.

**Pull request merged**